Add replace attribution to context menu and audit view

jonas-tai opened the pull request:
### Summary of changes

This commit adds the replace attribution function to the context menu and to the hamburger menu in audit view.

### Context and reason for change

Replacing attributions was only possible in attribution view so far and not available in the context menu. This adds these functionalities and allows the user to replace attributions with the context of the file tree and to switch between views while replacing attributions.

### How can the changes be tested

Run the context menu integration tests (part of the frontend ci integration tests) and / or test replace attributions with example file with context menu and in audit view.

leslielazzarino commented on src/Frontend/Components/AttributionColumn/AttributionColumn.tsx:
```diff
@@ -118,13 +116,21 @@ export function AttributionColumn(props: AttributionColumnProps): ReactElement {
     packageInfoWereModified,
-    Boolean(temporaryPackageInfo.preSelected)
+    Boolean(temporaryPackageInfo.preSelected),
+    false
```
This function has a huge number of parameters, and I do not see a meaningful way of breaking the function into parts. To make the code more readable, I would suggest the following js/ts pattern: the function takes a single object as input, where the keys are the names of the variables. Therefore when you are using the function you can see which input is what.

jonas-tai replied:
I changed the paramters of the function to a single object👍

antonbauhofer commented:
When selecting replace attribution from the context menu, the popup says `...to the current attribution`. This might be misleading, since the `current attribution` might not be the one selected (see screenshot). So, I suggest to remove `current` from the popup tex
<img width="1440" alt="Screenshot 2021-11-18 at 09 39 00" src="https://user-images.githubusercontent.com/48415604/142380767-b0fade71-7341-406c-9959-13c013a6362c.png">
t.

antonbauhofer commented on src/Frontend/Components/ResourceDetailsTabs/ResourceDetailsTabs.tsx:
```diff
@@ -73,12 +73,17 @@ export function ResourceDetailsTabs(
+      manualData is excluded from dependencies on purpose to avoid recalculation when
+      it changes. Usually this is not an issue as the displayed data remains correct.
+      In consequence, the panelData is eventually consistent.
+      Only manualData.attributionsToResources in dependencies to update panelData, when replaceAttributionPopup
+      was called in AuditView for manual attributions in attributions in folder content panel, but not update them
+      after every save click.
```
This sentence is very long and hard to read. Can you break it down into simpler sentences?

antonbauhofer commented on src/Frontend/Components/AttributionColumn/attribution-column-helpers.ts:
```diff
@@ -160,29 +160,39 @@ interface MergeButtonDisplayState {
+  isExternalAttribution: boolean
 ): MergeButtonDisplayState {
-  const isAttributionView = currentView === View.Attribution;
+  const isTableView = currentView === View.Report;
```
Let's call it `isReportView` to keep it consistent

jonas-tai replied:
Removed the parameter, because it is currently not necessary.

antonbauhofer reviewed (commented):
Looks good.
Just a few minor comments.

leslielazzarino commented on src/Frontend/Components/AttributionColumn/attribution-column-helpers.ts:
```diff
@@ -160,29 +160,39 @@ interface MergeButtonDisplayState {
+  isExternalAttribution: boolean
 ): MergeButtonDisplayState {
-  const isAttributionView = currentView === View.Attribution;
+  const isTableView = currentView === View.Report;
```
Is it at all possible to have the context menu on ReportView? I do not think so, as we have no cards there.

jonas-tai replied:
Removed the parameter, because it is currently not necessary👍

leslielazzarino commented on src/Frontend/Components/PackageCard/PackageCard.tsx:
```diff
@@ -259,6 +283,33 @@ export function PackageCard(props: PackageCardProps): ReactElement | null {
+              )
+            );
+          },
+          hidden: mergeButtonDisplayState.hideOnReplaceMarkedByButton,
```
Why "On" in the eneum name? I would remove it.

leslielazzarino commented on src/Frontend/Components/AttributionColumn/attribution-column-helpers.ts:
```diff
@@ -160,29 +160,39 @@ interface MergeButtonDisplayState {
       !anyAttributionMarkedForReplacement ||
-      !selectedAttributionIsMarkedForReplacement,
+      !targetAttributionIsMarkedForReplacement,
     hideOnReplaceMarkedByButton:
```
Why "On" in the name? I would remove it.

leslielazzarino commented on src/Frontend/integration-tests/all-views-tests/__tests-ci__/context-menu-all-views.test.tsx:
```diff
@@ -431,4 +434,71 @@ describe('The ContextMenu', () => {
     expectValueInTextBox(screen, 'Name', 'Vue');
   });
+
+  test('replace attributions buttons are synced with hamburger menu buttons', () => {
```
I do not think we need such an integration test. It is enough to test with unit tests that all buttons read/write correctly in the state. We do not need to test that Redux works with React.

jonas-tai replied:
Removed this test as part of the test refactoring.

leslielazzarino commented on src/Frontend/integration-tests/attribution-view-tests/__tests-ci__/context-menu-attribution-view.test.tsx:
```diff
@@ -139,4 +146,129 @@ describe('The ContextMenu', () => {
     expectValueInConfidenceField(screen, `High (${DiscreteConfidence.High})`);
   });
+
+  describe('replace attribution buttons', () => {
```
"replaces", the subject is the context menu. Same in the following row.

jonas-tai replied:
I think the whole naming was confusing to read, so I simplified the naming of these tests.

leslielazzarino commented on src/Frontend/integration-tests/audit-view-tests/__tests-ci__/context-menu-audit-view.test.tsx:
```diff
@@ -303,4 +294,195 @@ describe('The ContextMenu', () => {
     expectAddIconInAddToAttributionCardIsNotHidden(screen, 'Vue, 1.2.0');
   });
+
+  describe('replace attribution buttons', () => {
```
"replaces"

leslielazzarino commented on src/Frontend/integration-tests/audit-view-tests/__tests-ci__/context-menu-audit-view.test.tsx:
```diff
@@ -303,4 +294,195 @@ describe('The ContextMenu', () => {
+      '/root/src/file_2': ['uuid_2', 'uuid_3'],
+    };
+
+    test('replace attributions in Package panel', () => {
```
"replaces"

leslielazzarino reviewed (changes requested):
Too many integration tests. One should not test every bit of the feature through integration tests, as they are long to write/read and slow. My suggestion would be:
- a single integration test, that tests a complex enough case, e.g. that you can mark for replacement an attribution on attribution view and replace it with another on audit view
- everything else you can test through unit tests, where you just test that each component interacts correctly with the state, writing what it should or being updated correctly when something changes.

jonas-tai commented:
> Too many integration tests. One should not test every bit of the feature through integration tests, as they are long to write/read and slow. My suggestion would be:
> 
>     * a single integration test, that tests a complex enough case, e.g. that you can mark for replacement an attribution on attribution view and replace it with another on audit view
> 
>     * everything else you can test through unit tests, where you just test that each component interacts correctly with the state, writing what it should or being updated correctly when something changes.

Testing the replace attribution feature requires opening the corresponding replace attribution pop-up. The pop-up is opened at the top-level of the app, so it is not possible to test the feature in unit tests for each component. To make the integration tests easier to read, I removed the correct context menu configuration check from the integration tests. For that, I added separate unit tests for each component that verify the correct context menu configuration for the replace attribution buttons.
